File: Data Structures/Stack/Java/StackLinkedList.java
import java.util.*;

class Node {
    int data;
    Node next;
    
    Node(int data) {
        this.data = data;
        next = null;
    } 
}

class Stack {
    Node top;
    
    Stack() {
        top = null;
    }
    
    void push(int data) {
        Node newNode = new Node(data);
        
        if(top == null)
            top = newNode;
        else {
            newNode.next = top;
            top = newNode;
        }
    }
    
    int pop() {
        if(top == null)
            return 0;
        
        Node poppedNode = top;
        top = poppedNode.next;
        poppedNode.next = null;

        return poppedNode.data;
    }
    
    void display() {
        if(top == null)
            return;
            
        Node temp = top;
            
        while(temp != null) {
            System.out.print(temp.data + "->");
            temp = temp.next;
        }
            
        System.out.println("Null");
    }
}

public class StackLinkedList {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        Stack stack = new Stack();
        
        while (input.hasNextInt()) {
            stack.push(input.nextInt());
        }
        
        System.out.println("Stack contents are :");
        stack.display();

        System.out.println("Top element in Stack is : " + stack.top.data);
        System.out.println("Popped Element : " + stack.pop());
        
        input.close();
    }
}

/* Output:

1
2
3
4
5
\n
Stack contents are :
5->4->3->2->1->Null
Top element in Stack is : 5
Popped Element : 5

*/

File: Data Structures/Stack/Python/StackLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.top = None

    def isEmpty(self):
        if self.top == None:
            return True
        else:
            return False

    def push(self, data):
        newnode = Node(data)

        if self.isEmpty():
            self.top = newnode
        else:
            newnode.next = self.top
            self.top = newnode

    def pop(self):
        if self.isEmpty():
            return None
        else:
            poppednode = self.top
            self.top = self.top.next
            poppednode.next = None

            return poppednode.data

    def peek(self):
        if self.isEmpty():
            return None
        else:
            return self.top.data


MyStack = Stack()

MyStack.push(10)
MyStack.push(20)
MyStack.push(30)

print(MyStack.pop(), "popped from stack")
print("Top element is", MyStack.peek())


''' 
Output:

30 popped from stack
Top element is 20

'''
